### PR TITLE
Backwards compatibility for target platform compilation constants

### LINKED ISSUE #11236: Automatically define preprocessor symbols for target platforms
Spec is here: https://github.com/dotnet/designs/blob/master/accepted/2020/net5/net5.md#preprocessor-symbols

Note that this includes defining preprocessor symbols for the target platform, as well as defining preprocessor symbols for previous versions of .NET and the OS

### PR DESCRIPTION
**Pull request by sfoslund**
Fixes #11236

### CODE REVIEW

**Review by wli3 — APPROVED** [review]

**Comment by wli3** [comment]
Sorry for the delay

**Comment by sfoslund** [comment]
@dsplaisted I think I've made all the changes you suggested, let me know if there's anything I missed.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets [hunk 1]
```diff
@@ -206,10 +206,20 @@ Copyright (c) .NET Foundation. All rights reserved.
     </ItemGroup>
   </Target>
 
+  <!-- Add conditional compilation symbols for target platform backwards compatibility. -->
+  <Target Name="GeneratePlatformCompatibleDefineConstants"
+          Condition=" '$(DisableImplicitFrameworkDefines)' != 'true' and '$(TargetPlatformIdentifier)' != '' and '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and $([MSBuild]::VersionGreaterThanOrEquals($(TargetFrameworkVersion), 5.0)) " >
+    <ItemGroup>
+      <_SupportedPlatformCompatibleVersions Include="@(SupportedTargetPlatform)" Condition=" $(TargetPlatformMinVersion) != '' and %(Identity) != '' and $([MSBuild]::VersionGreaterThanOrEquals(%(Identity), $(TargetPlatformMinVersion))) and $([MSBuild]::VersionLessThan(%(Identity), $(TargetPlatformVersion))) " />
```
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets:213]: I'm not following what this logic is supposed to be doing with `TargetPlatformMinVersion`.  However, I think it should not be based on that property at all.  It should simply define constants for all supported versions less than or equal to the `TargetPlatformVersion`, pretty much the same way that it works for the `TargetFrameworkVersion`.

IE, I think that [this section of the spec](https://github.com/dotnet/designs/blob/master/accepted/2020/net5/net5.md#preprocessor-symbols) is only talking about `TargetPlatformVersion`, not the minimum version. @terrajobst @mhutch

**Review by dsplaisted — CHANGES_REQUESTED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets (hunk 1 above)
Comment (sfoslund) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets]: Sounds good, I've pushed a fix.

**Review by dsplaisted — APPROVED** [review]

**Comment by sfoslund** [comment]
@dsplaisted @wli3 This change hardcoded `SupportedTargetPlatform` into the test. Is there work being tracked that will add the real SupportedTargetPlatform item group?

**Comment by dsplaisted** [comment]
@sfoslund I've filed #12456 for adding `SupportedTargetPlatform` items (and generating an error if an invalid version is targeted).

**Pull request merged**